Backend/AlgorithmsManager.py

```diff
@@ -389,6 +389,67 @@ def parse_city_data(self, entry):
         city_data["race_multiple"] = entry[29]
 
         return city_data
+        
+    def get_user_by_username(self, username):
+        print("\nValidating user")
+        try:
+        # def __init__(self, id, ids, uname, signup, loc, protected, folcount, fricount):
+        # self.user_PK = id
+        # self.user_id = ids
+        # self.username = uname
+        # self.signup_date = signup
+        # self.location_id = loc
+        # self.protected = protected
+        # self.followers_count = folcount
+        # self.friends_count = fricount
+            ur = self.twitter.get_user(username)
+            user = DataStructures.User(None, ur.id, ur.screen_name, ur.name, ur.created_at, ur.location,
+                                        ur.protected, ur.followers_count, ur.friends_count)
+            res = []
+            res.append(user.__dict__)
+            print("\nValidation success")
+            return res
+        
+        except:
+            print("Validation fails")
+            return []
+
+    # this function returns a list of retweeters
+    #   args:   username
+    #           num_tweets (optional): how many recent tweets to search through
+    #           num_retweets (optional): how many retweeters per tweet
+    #           sorted (optional): value of 1 sorts the list by number of retweets 
+    #   return: a list of dictionary
+    def get_most_frequent_retweeters(self, username, num_tweets=20, num_retweets=20, sorted=1):
+
+        retweetData = self.twitter.most_recent_retweeters(username, num_tweets=num_tweets, num_retweets=num_retweets)
+        
+        # get the list of all retweeters
+        retweeter_history = []
+        for tweet in retweetData:
+            innerList = tweet['retweeters']
+            for username in innerList:
+                retweeter_history.append(username)
+
+        # count the number of retweets
+        res = []
+        for user_id in retweeter_history:
+            dict = {'username': user_id, 'retweets_count': 1}
+
+            if len(res) == 0:
+                res.append(dict)
```

Comment:
doesn't this only work for the first user_id in retweeter_history?

Reply:
Effectively yes.

Reply:
So I change the return object into a dictionary, no longer a list of dictionaries, which is much faster as you said. 
Sorting is done by the flask server with the `sort_retweeters` static method which sorts and shortens the result list at the same time (which is faster than sorting hundreds of items before cutting it off)